# BamView
BamView is an interactive Java application for visualising read-alignment data stored in BAM files.

## Publication
> So-called next-generation sequencing (NGS) has provided the ability to sequence on a massive scale at low cost, enabling biologists to perform powerful experiments and gain insight into biological processes. BamView has been developed to visualize and analyse sequence reads from NGS platforms, which have been aligned to a reference sequence. It is a desktop application for browsing the aligned or mapped reads [Ruffalo, M, LaFramboise, T, Koyutürk, M. Comparative analysis of algorithms for next-generation sequencing read alignment. Bioinformatics 2011;27:2790-6] at different levels of magnification, from nucleotide level, where the base qualities can be seen, to genome or chromosome level where overall coverage is shown. To enable in-depth investigation of NGS data, various views are provided that can be configured to highlight interesting aspects of the data. Multiple read alignment files can be overlaid to compare results from different experiments, and filters can be applied to facilitate the interpretation of the aligned reads. As well as being a standalone application it can be used as an integrated part of the Artemis genome browser, BamView allows the user to study NGS data in the context of the sequence and annotation of the reference genome. Single nucleotide polymorphism (SNP) density and candidate SNP sites can be highlighted and investigated, and read-pair information can be used to discover large structural insertions and deletions. The application will also calculate simple analyses of the read mapping, including reporting the read counts and reads per kilobase per million mapped reads (RPKM) for genes selected by the user.
   
If you make use of this software in your research, please cite as:
   
__BamView: visualizing and interpretation of next-generation sequencing read alignments.__   
Carver T, Harris SR, Otto TD, Berriman M, Parkhill J and McQuillan JA   
_Briefings in bioinformatics 2013;14;2;203-12_   
PUBMED: [22253280](http://ukpmc.ac.uk/abstract/MED/22253280); PMC: [3603209](http://ukpmc.ac.uk/articles/PMC3603209); DOI: [10.1093/bib/bbr073](http://dx.doi.org/10.1093/bib/bbr073)
  
__BamView: viewing mapped read alignment data in the context of the reference sequence.__  
Carver T, Böhme U, Otto TD, Parkhill J and Berriman M  
_Bioinformatics (Oxford, England) 2010;26;5;676-7_  
PUBMED: [20071372](http://ukpmc.ac.uk/abstract/MED/20071372); PMC: [2828118](http://ukpmc.ac.uk/articles/PMC2828118); DOI: [10.1093/bioinformatics/btq010](http://dx.doi.org/10.1093/bioinformatics/btq010)

## Software Availability
BamView is available under GPL3. The source code is available as part of the Artemis code which can be downloaded from [GitHub](https://github.com/sanger-pathogens/Artemis).
  
The latest release of BamView can be downloaded from our ftp site:

* [UNIX](ftp://ftp.sanger.ac.uk/pub/resources/software/artemis/bamview.jar)
* [MacOS](ftp://ftp.sanger.ac.uk/pub/resources/software/artemis/artemis.dmg.gz)
* [Windows](ftp://ftp.sanger.ac.uk/pub/resources/software/artemis/bamview.jar)

For further instructions please see the [BamView SourceForge page](http://bamview.sourceforge.net/).

## Installation 
For installation instructions please see the [BamView SourceForge page](http://bamview.sourceforge.net/).

## Usage
For instructions please see the [BamView SourceForge page](http://bamview.sourceforge.net/).

## Contact
For issues encountered with installing the software please contact your local system administrator. For all other issues, please report them to our [Github issues page](https://github.com/sanger-pathogens/Artemis/issues) or email <artemis-help@sanger.ac.uk>.
