# Artemis Comparison Tool (ACT)
ACT is a Java application for displaying pairwise comparisons between two or more DNA sequences. It can be used to identify and analyse regions of similarity and difference between genomes and to explore conservation of synteny, in the context of the entire sequences and their annotation. It can read complete EMBL, GENBANK and GFF entries or sequences in FASTA or raw format.

### Related Software
Two separate ACT-related web applications have been developed.

* [WebACT](http://www.webact.org/WebACT/home) was developed at Imperial College, and allows users to generate comparisons from their sequence files or use a set of pre-computed comparisons. All comparison files can be downloaded for local use or launched in ACT.
* [DoubleACT](http://www.hpa-bioinfotools.org.uk/pise/double_actv2.html) was developed by Anthony Underwood and Jonathan Green at the [Public Public Health England](http://www.hpa.org.uk/) and allows you to paste or upload sequences to generate ACT comparison files.

## Publication
> The Artemis Comparison Tool (ACT) allows an interactive visualisation of comparisons between complete genome sequences and associated annotations. The comparison data can be generated with several different programs; BLASTN, TBLASTX or Mummer comparisons between genomic DNA sequences, or orthologue tables generated by reciprocal FASTA comparison between protein sets. It is possible to identify regions of similarity, insertions and rearrangements at any level from the whole genome to base-pair differences. ACT uses Artemis components to display the sequences and so inherits powerful searching and analysis tools. ACT is part of the Artemis distribution and is similarly open source, written in Java and can run on any Java enabled platform, including UNIX, Macintosh and Windows.

If you make use of this software in your research, please cite as:

__ACT: the Artemis Comparison Tool.__  
Carver TJ, Rutherford KM, Berriman M, Rajandream MA, Barrell BG and Parkhill  
_Bioinformatics (Oxford, England) 2005;21;16;3422-3_   
PUBMED: [15976072](http://ukpmc.ac.uk/abstract/MED/15976072); DOI: [10.1093/bioinformatics/bti553](http://dx.doi.org/10.1093/bioinformatics/bti553)

## Software Availability
The Artemis Software is available under GPL3. The source code can be found on [GitHub](https://github.com/sanger-pathogens/Artemis).

The latest release of Artemis can be downloaded from our ftp site:

* [UNIX](ftp://ftp.sanger.ac.uk/pub/resources/software/artemis/artemis.tar.gz)
* [MacOS](ftp://ftp.sanger.ac.uk/pub/resources/software/artemis/artemis.dmg.gz)
* [Windows](ftp://ftp.sanger.ac.uk/pub/resources/software/artemis/artemis-windows.zip)

For older versions of the Artemis software please see the [Artemis FTP site](ftp://ftp.sanger.ac.uk/pub/resources/software/artemis/)

### Notes for Download

__Note for UNIX/Windows operating systems__: The Artemis tool suite requires Java version 1.8 to run. This must be installed first.

__Note for MacOSX__: occasionally a browser decides to display the contents of the .dmg.gz archive file rather than downloading it. If this happens hold down the <control> key and click on the download link. A popup menu should appear, containing several choices. One of the choices should be something like "Save Link As" (or perhaps "Download Link...", "Save Link to Desktop", or a variation on this theme). Select that option, and the archive file should be download correctly.

## Installation
### For UNIX/Linux
Copy the artemis.tar.gz file to the directory that you wish to install to. Change to that directory and uncompress and untar the file. On UNIX, the command is:
```
tar zxf artemis.tar.gz
```
### For MacOSX
For MacOSX users an archive artemis.dmg.gz disk image is provided. Click on this file in your Downloads folder. This should automatically unzip and mount the disk image, opening a Finder window that contains the Artemis, ACT, BamView and DNAPlotter applications. These apps can then be dragged to any desired location, for example, your desktop.

### For Windows
Copy the artemis-windows.zip file to the directory that you wish to install to and then unzip using an application such as WinZip.
This should unpack the artemis.jar, act.jar, bamview.jar and dnaplotter.jar application files.

## Usage
### Running ACT on UNIX/Linux Systems
The easiest way to run the program is to run the script called act in the Artemis installation directory, like this:
```
artemis/act
```
Alternatively you can start ACT by passing the comparison filenames as arguments, eg:
```
artemis/act sequence1 blast_output.crunch sequence2
```
For other available options use:
```
artemis/act -help
```
### Running ACT on Macintosh Systems
On MacOSX machines, ACT can be started by double clicking on the ACT icon.

### Running ACT on Windows Systems
ACT can be started by double clicking on the act.jar icon.

__Note__: For additional information please see the [ACT manual](https://sanger-pathogens.github.io/Artemis/ACT/act-manual.html). A PDF version of the manual is also available for download [here](https://sanger-pathogens.github.io/Artemis/ACT/act-manual.pdf).

## Contact
For issues encountered with installing the software please contact your local system administrator. For all other issues, please report them to our [Github issues page](https://github.com/sanger-pathogens/Artemis/issues) or email <artemis-help@sanger.ac.uk>.

## FAQ
### Why does ACT run out of memory on UNIX or GNU/Linux even though the machine has lots of memory?

The Java Virtual Machine (JVM) on UNIX has a fixed upper limit on the amount of memory that is available for applications, but this limit can be changed at runtime. As shipped ACT will use a maximum of 2GB of memory.

There are two ways of fixing this problem:
1. Change the act script. Find the line that reads: FLAGS="-mx2g -ms100m -noverify" and change the 2g (2 gigabytes) to a bigger number dependent on your machine memory (try 3g), or
2. Create an ARTEMIS_JVM_FLAGS environment variable set to "-mx2g -ms100m", adjusting the mx value as required. No script change is required for this, but it would need to be added to your environment.

### Why does ACT run out of memory on MacOSX even though the machine has lots of memory?
To change the memory allocated to ACT on MacOSX, set the value in the file ACT.cfg in the directory ACT.app/Contents/Java. There are a couple of lines that look like this:

```
[JVMOptions]
-Xmx2g
```
Changing the value after -Xmx will change the memory used by ACT.

### Why does ACT run out of memory on Windows even though the machine has lots of memory?

Normally the Java virtual machine artificially limits the amount of memory that ACT can use. The fix is as follows:

Create a shortcut to the act.jar JAR file. Edit the properties of the shortcut and add java -mx2g -jar to the start of the Target: field. -mx2g sets the maximum memory Java will allocate to ACT (2 gigabytes in this case). We recommend choosing a number that is about 50 megabytes less than the total amount of memory in the machine (to allow for the overhead of windows and the Java virtual machine).

You will need to use the shortcut to run ACT from then on.

## References
__WebACT: an online genome comparison suite.__   
Abbott JC, Aanensen DM and Bentley SD  
_Methods in molecular biology (Clifton, N.J.) 2007;395;57-74_  
PUBMED: [17993667](http://ukpmc.ac.uk/abstract/MED/17993667)
