/**
 * Scaffolding file used to store all the setups needed to run 
 * tests automatically generated by EvoSuite
 * Fri Jan 12 11:15:39 GMT 2018
 */

package uk.ac.sanger.artemis.components.alignment;

import org.evosuite.runtime.annotation.EvoSuiteClassExclude;
import org.junit.BeforeClass;
import org.junit.Before;
import org.junit.After;
import org.junit.AfterClass;
import org.evosuite.runtime.sandbox.Sandbox;
import org.evosuite.runtime.sandbox.Sandbox.SandboxMode;

import static org.evosuite.shaded.org.mockito.Mockito.*;
@EvoSuiteClassExclude
public class BamUtils_ESTest_scaffolding {

  @org.junit.Rule 
  public org.evosuite.runtime.vnet.NonFunctionalRequirementRule nfr = new org.evosuite.runtime.vnet.NonFunctionalRequirementRule();

  private static final java.util.Properties defaultProperties = (java.util.Properties) java.lang.System.getProperties().clone(); 

  private org.evosuite.runtime.thread.ThreadStopper threadStopper =  new org.evosuite.runtime.thread.ThreadStopper (org.evosuite.runtime.thread.KillSwitchHandler.getInstance(), 3000);


  @BeforeClass 
  public static void initEvoSuiteFramework() { 
    org.evosuite.runtime.RuntimeSettings.className = "uk.ac.sanger.artemis.components.alignment.BamUtils"; 
    org.evosuite.runtime.GuiSupport.initialize(); 
    org.evosuite.runtime.RuntimeSettings.maxNumberOfThreads = 100; 
    org.evosuite.runtime.RuntimeSettings.maxNumberOfIterationsPerLoop = 10000; 
    org.evosuite.runtime.RuntimeSettings.mockSystemIn = true; 
    org.evosuite.runtime.RuntimeSettings.sandboxMode = org.evosuite.runtime.sandbox.Sandbox.SandboxMode.RECOMMENDED; 
    org.evosuite.runtime.sandbox.Sandbox.initializeSecurityManagerForSUT(); 
    org.evosuite.runtime.classhandling.JDKClassResetter.init();
    setSystemProperties();
    initializeClasses();
    org.evosuite.runtime.Runtime.getInstance().resetRuntime(); 
    try { initMocksToAvoidTimeoutsInTheTests(); } catch(ClassNotFoundException e) {} 
  } 

  @AfterClass 
  public static void clearEvoSuiteFramework(){ 
    Sandbox.resetDefaultSecurityManager(); 
    java.lang.System.setProperties((java.util.Properties) defaultProperties.clone()); 
  } 

  @Before 
  public void initTestCase(){ 
    threadStopper.storeCurrentThreads();
    threadStopper.startRecordingTime();
    org.evosuite.runtime.jvm.ShutdownHookHandler.getInstance().initHandler(); 
    org.evosuite.runtime.sandbox.Sandbox.goingToExecuteSUTCode(); 
    setSystemProperties(); 
    org.evosuite.runtime.GuiSupport.setHeadless(); 
    org.evosuite.runtime.Runtime.getInstance().resetRuntime(); 
    org.evosuite.runtime.agent.InstrumentingAgent.activate(); 
  } 

  @After 
  public void doneWithTestCase(){ 
    threadStopper.killAndJoinClientThreads();
    org.evosuite.runtime.jvm.ShutdownHookHandler.getInstance().safeExecuteAddedHooks(); 
    org.evosuite.runtime.classhandling.JDKClassResetter.reset(); 
    resetClasses(); 
    org.evosuite.runtime.sandbox.Sandbox.doneWithExecutingSUTCode(); 
    org.evosuite.runtime.agent.InstrumentingAgent.deactivate(); 
    org.evosuite.runtime.GuiSupport.restoreHeadlessMode(); 
  } 

  public static void setSystemProperties() {
 
    java.lang.System.setProperties((java.util.Properties) defaultProperties.clone()); 
    java.lang.System.setProperty("file.encoding", "UTF-8"); 
    java.lang.System.setProperty("java.awt.headless", "true"); 
    java.lang.System.setProperty("user.country", "GB"); 
    java.lang.System.setProperty("user.language", "en"); 
    java.lang.System.setProperty("user.timezone", "Europe/London"); 
    java.lang.System.setProperty("log4j.configuration", "SUT.log4j.properties"); 
    java.lang.System.setProperty("sun.jnu.encoding", "UTF-8"); 
  }

  private static void initializeClasses() {
    org.evosuite.runtime.classhandling.ClassStateSupport.initializeClasses(BamUtils_ESTest_scaffolding.class.getClassLoader() ,
      "uk.ac.sanger.artemis.components.LogViewer$1",
      "htsjdk.samtools.cram.structure.ContainerIO",
      "org.biojava.bio.seq.AcceptNoneFilter",
      "uk.ac.sanger.artemis.components.alignment.AbstractGraphPanel",
      "htsjdk.samtools.SamFileValidator$MaxOutputExceededException",
      "org.biojava.bio.seq.io.SymbolTokenization$TokenType",
      "htsjdk.samtools.SamFiles",
      "org.biojava.bio.seq.ByHierarchy",
      "org.apache.log4j.AppenderSkeleton",
      "uk.ac.sanger.artemis.OptionChangeEvent",
      "uk.ac.sanger.artemis.components.genebuilder.JExtendedComboBox",
      "org.biojava.utils.stax.SAX2StAXAdaptor",
      "uk.ac.sanger.artemis.components.EntryEdit$CommitButton",
      "uk.ac.sanger.artemis.components.variant.VCFview$27",
      "uk.ac.sanger.artemis.components.ListSelectionPanel",
      "uk.ac.sanger.artemis.io.SimpleDocumentFeature",
      "uk.ac.sanger.artemis.sequence.BasePatternFormatException",
      "uk.ac.sanger.artemis.components.BasePlotGroup",
      "htsjdk.samtools.util.SequenceUtil$SequenceListsDifferException",
      "com.sshtools.j2ssh.authentication.PasswordAuthenticationClient",
      "uk.ac.sanger.artemis.components.variant.VCFview",
      "uk.ac.sanger.artemis.plot.CodonUsageAlgorithm",
      "com.sshtools.j2ssh.SshClient",
      "uk.ac.sanger.artemis.EntryChangeListener",
      "org.biojava.utils.ListTools$Triplet",
      "uk.ac.sanger.artemis.sequence.AminoAcidSequence",
      "org.biojava.bio.symbol.DoubleAlphabet$SubDoubleAlphabet",
      "org.biojava.bio.symbol.AlphabetManager$AlphabetManagerHandler$AlphabetHandler$1",
      "uk.ac.sanger.artemis.Selectable",
      "org.biojava.bio.seq.io.NameTokenization",
      "org.apache.log4j.spi.Filter",
      "uk.ac.sanger.artemis.io.FeatureTable",
      "htsjdk.samtools.reference.ReferenceSequenceFileFactory",
      "htsjdk.variant.vcf.VCFIDHeaderLine",
      "uk.ac.sanger.artemis.io.RangeVector",
      "org.biojava.bio.seq.FramedFeature$ReadingFrame",
      "uk.ac.sanger.artemis.plot.NcAlgorithm",
      "org.biojava.bio.seq.io.CharacterTokenization",
      "org.biojava.bio.seq.homol.SimilarityPairFeature$EmptyPairwiseAlignment",
      "htsjdk.samtools.util.AsyncBlockCompressedInputStream$1",
      "uk.ac.sanger.artemis.io.PublicDBDocumentEntry",
      "htsjdk.samtools.reference.ReferenceSequenceFileFactory$1",
      "uk.ac.sanger.artemis.util.StringVector",
      "org.emboss.jemboss.JembossParams",
      "htsjdk.samtools.util.CloseableIterator",
      "htsjdk.samtools.AbstractSAMHeaderRecord",
      "org.biojava.bio.seq.Feature$Template",
      "com.sshtools.j2ssh.transport.AsyncService",
      "uk.ac.sanger.artemis.plot.CumulativeATSkewAlgorithm",
      "htsjdk.samtools.SamFileValidator",
      "org.biojava.bio.symbol.SimpleAlphabet",
      "htsjdk.samtools.CRAMFileReader",
      "htsjdk.samtools.SAMFileHeader$SortOrder",
      "uk.ac.sanger.artemis.components.genebuilder.gff.RemoveButton",
      "uk.ac.sanger.artemis.components.genebuilder.gff.PropertiesPanel",
      "org.apache.log4j.Hierarchy",
      "htsjdk.samtools.seekablestream.SeekableFileStream",
      "uk.ac.sanger.artemis.sequence.MarkerRange",
      "org.biojava.utils.Changeable",
      "htsjdk.samtools.util.CustomGzipOutputStream",
      "uk.ac.sanger.artemis.plot.GCFrameAlgorithm",
      "uk.ac.sanger.artemis.components.genebuilder.ReferencesPanel",
      "uk.ac.sanger.artemis.io.QualifierInfoHash",
      "com.sshtools.j2ssh.connection.Channel",
      "uk.ac.sanger.artemis.SimpleEntryGroup",
      "org.apache.log4j.spi.OptionHandler",
      "org.biojava.bio.symbol.SimpleManyToOneTranslationTable",
      "htsjdk.samtools.util.DateParser$InvalidDateException",
      "org.biojava.utils.stax.DelegationManager",
      "uk.ac.sanger.artemis.plot.PositionalAsymmetryAlgorithm",
      "org.biojava.bio.seq.StrandedFeature",
      "org.biojava.utils.lsid.LifeScienceIdentifier",
      "org.biojava.utils.ChangeVetoException",
      "org.biojava.bio.symbol.IllegalAlphabetException",
      "htsjdk.samtools.util.BufferedLineReader",
      "org.biojava.bio.symbol.DoubleAlphabet$DoubleRange",
      "htsjdk.samtools.seekablestream.SeekableStream",
      "htsjdk.samtools.DiskBasedBAMFileIndex",
      "htsjdk.samtools.QueryInterval",
      "com.sshtools.j2ssh.connection.ChannelEventListener",
      "uk.ac.sanger.artemis.components.Splash$2",
      "org.biojava.bio.symbol.AbstractTranslationTable",
      "org.gmod.schema.dao.GeneralDaoI",
      "org.apache.log4j.helpers.OptionConverter",
      "org.gmod.schema.dao.BaseDaoI",
      "org.biojava.bio.seq.SimpleFeatureRealizer$TemplateImpl",
      "org.biojava.bio.symbol.Edit",
      "htsjdk.samtools.FileInputResource$1",
      "uk.ac.sanger.artemis.io.MiscLineGroup",
      "uk.ac.sanger.artemis.circular.TextFieldInt",
      "uk.ac.sanger.artemis.editor.MultiLineToolTipUI",
      "htsjdk.samtools.util.Histogram",
      "uk.ac.sanger.artemis.io.LazyQualifierValue",
      "org.biojava.bio.symbol.AbstractSymbol",
      "uk.ac.sanger.artemis.components.FeatureDisplay$11",
      "org.biojava.utils.bytecode.CodeGenerator",
      "org.gmod.schema.cv.CvTerm",
      "htsjdk.samtools.SAMFileHeader",
      "uk.ac.sanger.artemis.sequence.SequenceChangeListener",
      "uk.ac.sanger.artemis.io.InvalidRelationException",
      "org.biojava.utils.bytecode.CodeMethod",
      "htsjdk.samtools.AbstractBAMFileIndex",
      "org.biojava.bio.seq.impl.FeatureImpl$1",
      "com.sshtools.j2ssh.transport.InvalidMessageException",
      "uk.ac.sanger.artemis.components.genebuilder.cv.CVPanel",
      "org.biojava.bio.Annotation",
      "org.biojava.bio.symbol.IntegerAlphabet$IntegerSymbol",
      "org.biojava.utils.lsid.LifeScienceIdentifierParseException",
      "org.biojava.bio.symbol.AlphabetManager$AlphabetManagerHandler$CharacterTokenizationHandler",
      "com.sshtools.j2ssh.authentication.AuthenticationProtocolClient",
      "org.apache.log4j.Category",
      "htsjdk.samtools.FileInputResource",
      "uk.ac.sanger.artemis.io.LocationParseException",
      "htsjdk.samtools.cram.structure.ContainerIO$1",
      "org.biojava.utils.ListTools$Mapper",
      "htsjdk.samtools.SAMException",
      "htsjdk.samtools.SAMTextReader",
      "org.biojava.bio.seq.io.StreamParser",
      "htsjdk.samtools.SAMBinaryTagAndValue",
      "uk.ac.sanger.artemis.util.FastVector",
      "org.biojava.bio.symbol.SimpleSymbol",
      "htsjdk.samtools.seekablestream.SeekableFTPStream",
      "org.apache.log4j.helpers.ThreadLocalMap",
      "org.biojava.bio.symbol.AbstractSimpleBasisSymbol",
      "htsjdk.samtools.SAMRecord",
      "htsjdk.samtools.SAMTextHeaderCodec$HeaderRecordType",
      "com.sshtools.j2ssh.transport.TransportProtocolEventHandler",
      "org.gmod.schema.sequence.FeatureDbXRef",
      "htsjdk.samtools.SAMProgramRecord",
      "org.biojava.utils.ChangeEvent",
      "uk.ac.sanger.artemis.util.URLDocument",
      "htsjdk.samtools.SamReaderFactory$SamReaderFactoryImpl",
      "org.gmod.schema.pub.Pub",
      "htsjdk.samtools.SamReader$PrimitiveSamReaderToSamReaderAdapter",
      "uk.ac.sanger.artemis.io.ChadoCanonicalGene",
      "org.biojava.utils.stax.SAX2StAXAdaptor$S2SDelegationManager",
      "ngs.ErrorMsg",
      "htsjdk.tribble.util.RemoteURLHelper",
      "uk.ac.sanger.artemis.components.alignment.BamView$41",
      "org.biojava.bio.symbol.SimpleCrossProductAlphabet",
      "htsjdk.samtools.SamReaderFactory$Option",
      "uk.ac.sanger.artemis.components.Plot",
      "org.biojava.bio.seq.DNATools",
      "com.sshtools.j2ssh.transport.TransportProtocolState",
      "htsjdk.samtools.util.Lazy",
      "org.biojava.bio.symbol.AbstractManyToOneTranslationTable",
      "uk.ac.sanger.artemis.EntrySource",
      "uk.ac.sanger.artemis.io.GenbankStreamSequence",
      "uk.ac.sanger.artemis.components.alignment.BamView$47",
      "htsjdk.samtools.cram.io.ExposedByteArrayOutputStream",
      "htsjdk.samtools.SAMTextHeaderCodec",
      "uk.ac.sanger.artemis.components.SegmentBorder",
      "com.sshtools.j2ssh.io.UnsignedInteger32",
      "htsjdk.samtools.util.BlockCompressedInputStream$DecompressedBlock",
      "uk.ac.sanger.artemis.plot.CodonUsageWeight",
      "htsjdk.samtools.util.RuntimeIOException",
      "htsjdk.samtools.InputResource",
      "org.biojava.bio.BioRuntimeException",
      "htsjdk.samtools.SAMRecordFactory",
      "org.biojava.bio.seq.AcceptAllFilter",
      "org.biojava.bio.symbol.IntegerAlphabet$SubIntegerAlphabet",
      "htsjdk.samtools.SamReader$Type$TypeImpl",
      "org.biojava.bio.seq.homol.SimilarityPairFeature",
      "htsjdk.samtools.util.Locatable",
      "uk.ac.sanger.artemis.plot.KarlinSigAlgorithm",
      "org.gmod.schema.utils.propinterface.PropertyI",
      "uk.ac.sanger.artemis.plot.EntropyAlgorithm",
      "uk.ac.sanger.artemis.components.DisplayAdjustmentEvent",
      "htsjdk.samtools.SAMSequenceDictionary",
      "org.apache.log4j.spi.AppenderAttachable",
      "com.sshtools.j2ssh.SshException",
      "htsjdk.samtools.util.zip.InflaterFactory",
      "uk.ac.sanger.artemis.j2ssh.FTProgress",
      "uk.ac.sanger.artemis.io.GFF3AttributeAggregator",
      "uk.ac.sanger.artemis.util.InputStreamProgressListenerVector",
      "uk.ac.sanger.artemis.util.PushBackException",
      "uk.ac.sanger.artemis.components.FeatureDisplay$12",
      "com.sshtools.j2ssh.connection.IOChannel",
      "org.biojava.utils.SingletonList",
      "org.biojava.bio.symbol.SymbolList",
      "org.biojava.bio.BioException",
      "uk.ac.sanger.artemis.plot.CodonUsageFormatException",
      "htsjdk.samtools.CachingBAMFileIndex",
      "org.biojava.bio.seq.FramedFeature$Template",
      "uk.ac.sanger.artemis.io.DatabaseDocumentEntry",
      "htsjdk.samtools.seekablestream.SeekableStreamFactory$DefaultSeekableStreamFactory",
      "uk.ac.sanger.artemis.io.LocationParseNode",
      "uk.ac.sanger.artemis.plot.Codon12CorrelationAlgorithm",
      "htsjdk.samtools.AbstractBAMFileIndex$MemoryMappedFileBuffer",
      "uk.ac.sanger.artemis.util.WorkingGZIPInputStream",
      "htsjdk.samtools.reference.ReferenceSequenceFile",
      "org.biojava.utils.ChangeType$1",
      "org.biojava.bio.symbol.AlphabetManager$WellKnownAlphabet",
      "htsjdk.variant.vcf.VCFCompoundHeaderLine",
      "uk.ac.sanger.artemis.components.EditMenu",
      "uk.ac.sanger.artemis.io.ComparableFeature",
      "uk.ac.sanger.artemis.SelectionChangeEvent",
      "uk.ac.sanger.artemis.plot.BaseAlgorithm",
      "uk.ac.sanger.artemis.io.Key",
      "uk.ac.sanger.artemis.Options",
      "org.biojava.bio.seq.impl.SimpleHomologyFeature",
      "uk.ac.sanger.artemis.io.SimpleDocumentEntry",
      "htsjdk.variant.vcf.VCFSimpleHeaderLine",
      "htsjdk.samtools.SAMRecordIterator",
      "org.biojava.bio.symbol.BasisSymbol",
      "uk.ac.sanger.artemis.io.FeatureHeader",
      "org.biojava.bio.seq.RNATools",
      "htsjdk.samtools.CustomReaderFactory",
      "htsjdk.samtools.util.IOUtil$2",
      "uk.ac.sanger.artemis.components.filetree.FileList",
      "com.sshtools.j2ssh.connection.ConnectionProtocol",
      "htsjdk.samtools.SamReader$Indexing",
      "htsjdk.samtools.SAMFlag",
      "uk.ac.sanger.artemis.plot.ATDeviationAlgorithm",
      "org.biojava.bio.seq.io.WordTokenization",
      "uk.ac.sanger.artemis.ChangeListener",
      "uk.ac.sanger.artemis.io.FeatureTree",
      "uk.ac.sanger.artemis.components.MarkerRangeRequesterListener",
      "uk.ac.sanger.artemis.plot.LineAttributes",
      "uk.ac.sanger.artemis.components.alignment.BamFrame",
      "htsjdk.samtools.BAMFileReader",
      "org.apache.log4j.helpers.Loader",
      "uk.ac.sanger.artemis.sequence.MarkerChangeEvent",
      "org.biojava.utils.stax.SAX2StAXAdaptor$HandlerBinding",
      "htsjdk.samtools.SamReader$Type",
      "htsjdk.samtools.cram.ref.ReferenceSource",
      "org.biojava.bio.seq.impl.SimpleFeature",
      "org.apache.log4j.helpers.FileWatchdog",
      "uk.ac.sanger.artemis.components.genebuilder.GeneEditorPanel",
      "uk.ac.sanger.artemis.components.SelectionMenu",
      "uk.ac.sanger.artemis.components.KeyChoice",
      "org.biojava.bio.symbol.AbstractSymbolList$ListView",
      "uk.ac.sanger.artemis.io.RawStreamSequence",
      "uk.ac.sanger.artemis.components.alignment.BamUtils",
      "htsjdk.samtools.util.CollectionUtil",
      "org.biojava.bio.seq.GappedSequence",
      "org.biojava.bio.symbol.Alignment",
      "org.apache.log4j.MDC",
      "org.biojava.bio.seq.impl.SimpleRestrictionSite",
      "htsjdk.samtools.CustomReaderFactory$ICustomReaderFactory",
      "uk.ac.sanger.artemis.components.alignment.SAMRecordPredicate",
      "uk.ac.sanger.artemis.io.QualifierInfoException",
      "org.gmod.schema.dao.SequenceDaoI",
      "uk.ac.sanger.artemis.sequence.SequenceChangeEvent",
      "htsjdk.samtools.cram.structure.Container",
      "htsjdk.samtools.SeekableStreamInputResource",
      "org.biojava.utils.SmallMap",
      "org.biojava.bio.seq.StrandedFeature$Template",
      "htsjdk.samtools.cram.structure.BlockContentType",
      "htsjdk.samtools.CRAMCRAIIndexer",
      "com.sshtools.j2ssh.transport.HostKeyVerification",
      "htsjdk.samtools.BAMFileSpan",
      "org.biojava.bio.symbol.Symbol",
      "htsjdk.samtools.SAMFileHeader$GroupOrder",
      "org.biojava.bio.seq.RealizingFeatureHolder",
      "org.biojava.bio.molbio.RestrictionSite$Template",
      "com.sshtools.j2ssh.io.UnsignedInteger64",
      "htsjdk.samtools.BAMIndex",
      "org.gmod.schema.sequence.FeatureCvTerm",
      "htsjdk.samtools.SamReaderFactory$Option$2",
      "org.biojava.bio.seq.FramedFeature",
      "htsjdk.samtools.SamReaderFactory$Option$1",
      "htsjdk.samtools.SamReaderFactory$Option$4",
      "htsjdk.samtools.SamReaderFactory$Option$3",
      "uk.ac.sanger.artemis.components.alignment.CoveragePanel",
      "htsjdk.samtools.SamReaderFactory$Option$5",
      "org.biojava.utils.ListTools",
      "uk.ac.sanger.artemis.components.genebuilder.gff.AddButton",
      "uk.ac.sanger.artemis.components.alignment.LookSeqPanel$13",
      "org.biojava.bio.symbol.AlphabetManager$AlphabetManagerHandler$AlphabetHandler",
      "uk.ac.sanger.artemis.io.DatabaseStreamFeature",
      "org.gmod.schema.dao.CvDaoI",
      "uk.ac.sanger.artemis.plot.GCWindowAlgorithm",
      "htsjdk.tribble.TribbleException",
      "org.biojava.bio.seq.Feature",
      "uk.ac.sanger.artemis.util.FTPSeekableStream",
      "uk.ac.sanger.artemis.io.InvalidKeyException",
      "htsjdk.samtools.DefaultSAMRecordFactory",
      "uk.ac.sanger.artemis.circular.TextFieldSink",
      "uk.ac.sanger.artemis.io.Packing",
      "org.biojava.bio.symbol.AlphabetManager$ImmutableWellKnownAlphabetWrapper",
      "org.biojava.bio.seq.Sequence",
      "uk.ac.sanger.artemis.util.ByteBuffer",
      "htsjdk.samtools.SAMValidationError$Severity",
      "uk.ac.sanger.artemis.components.DisplayAdjustmentListener",
      "uk.ac.sanger.artemis.io.SimpleEntryInformation",
      "htsjdk.samtools.seekablestream.ISeekableStreamFactory",
      "htsjdk.samtools.cram.common.Version",
      "org.biojava.utils.stax.StAXContentHandlerBase",
      "uk.ac.sanger.artemis.components.ZoomScrollBar",
      "uk.ac.sanger.artemis.io.StreamSequence",
      "uk.ac.sanger.artemis.io.Range",
      "uk.ac.sanger.artemis.sequence.Bases",
      "uk.ac.sanger.artemis.EntryGroupChangeListener",
      "uk.ac.sanger.artemis.components.LogViewer",
      "uk.ac.sanger.artemis.ActionController",
      "uk.ac.sanger.artemis.io.BetweenRange",
      "uk.ac.sanger.artemis.ExternalProgramMonitor",
      "htsjdk.samtools.util.Iso8601Date",
      "com.sshtools.j2ssh.transport.SshMessage",
      "htsjdk.samtools.Cigar",
      "htsjdk.samtools.util.RuntimeEOFException",
      "com.sshtools.j2ssh.util.State",
      "uk.ac.sanger.artemis.j2ssh.SshLogin",
      "org.biojava.bio.EmptyAnnotation",
      "uk.ac.sanger.artemis.plot.CumulativeGCSkewAlgorithm",
      "org.apache.log4j.PropertyConfigurator",
      "htsjdk.tribble.TribbleException$InvalidHeader",
      "uk.ac.sanger.artemis.util.Document",
      "htsjdk.samtools.util.IOUtil$DeletePathThread",
      "htsjdk.samtools.ValidationStringency",
      "org.biojava.bio.symbol.AlphabetManager",
      "htsjdk.samtools.SAMTag",
      "htsjdk.samtools.BrowseableBAMIndex",
      "uk.ac.sanger.artemis.components.FileViewer",
      "uk.ac.sanger.artemis.Selection",
      "htsjdk.samtools.util.Histogram$Bin",
      "org.biojava.utils.walker.WalkerFactory",
      "org.biojava.bio.seq.FeatureRealizer",
      "org.biojava.utils.bytecode.CodeException",
      "uk.ac.sanger.artemis.sequence.NoSequenceException",
      "htsjdk.variant.vcf.VCFFilterHeaderLine",
      "org.biojava.bio.symbol.SparseCrossProductAlphabet",
      "uk.ac.sanger.artemis.components.QualifierTextArea",
      "uk.ac.sanger.artemis.components.SelectionInfoDisplay",
      "uk.ac.sanger.artemis.plot.ICDIAlgorithm",
      "htsjdk.samtools.cram.structure.SliceIO$1",
      "org.biojava.bio.symbol.AlphabetManager$WellKnownBasisSymbol",
      "org.biojava.bio.seq.impl.SimpleRemoteFeature",
      "htsjdk.samtools.SAMRecordComparator",
      "htsjdk.samtools.AbstractBAMFileIndex$IndexStreamBuffer",
      "org.apache.log4j.spi.LoggerRepository",
      "com.sshtools.j2ssh.transport.TransportProtocolException",
      "htsjdk.samtools.SAMSequenceRecord",
      "uk.ac.sanger.artemis.components.Splash",
      "uk.ac.sanger.artemis.components.TextRequesterListener",
      "uk.ac.sanger.artemis.io.EntryInformationException",
      "com.sshtools.j2ssh.authentication.SshMsgUserAuthRequest",
      "uk.ac.sanger.artemis.io.GFFMisc",
      "htsjdk.samtools.SAMRecordDuplicateComparator",
      "uk.ac.sanger.artemis.FeatureSegmentVector",
      "uk.ac.sanger.artemis.components.genebuilder.OpenSectionButton",
      "uk.ac.sanger.artemis.components.alignment.ReadCount",
      "uk.ac.sanger.artemis.components.AddMenu",
      "htsjdk.samtools.util.Log$LogLevel",
      "htsjdk.samtools.FileTruncatedException",
      "htsjdk.samtools.util.IterableOnceIterator",
      "htsjdk.samtools.SAMRecordQueryNameComparator",
      "uk.ac.sanger.artemis.chado.IBatisDAO",
      "uk.ac.sanger.artemis.components.filetree.RemoteFileNode",
      "org.biojava.bio.seq.impl.SimpleSimilarityPairFeature",
      "uk.ac.sanger.artemis.io.SimpleDocumentEntry$1",
      "org.biojava.bio.symbol.SimpleSymbolList$SSLIOListener",
      "org.biojava.bio.seq.RemoteFeature$Template",
      "uk.ac.sanger.artemis.components.alignment.SAMRecordMapQPredicate",
      "uk.ac.sanger.artemis.plot.GCDeviationAlgorithm",
      "org.apache.log4j.Level",
      "uk.ac.sanger.artemis.components.RunMenu",
      "org.biojava.bio.symbol.AlphabetManager$WellKnownAtomicSymbol",
      "org.biojava.bio.symbol.AbstractReversibleTranslationTable",
      "htsjdk.samtools.InputStreamInputResource",
      "org.biojava.bio.seq.io.SeqIOAdapter",
      "uk.ac.sanger.artemis.components.EntryGroupDisplay$3",
      "uk.ac.sanger.artemis.components.FeatureDisplay",
      "com.sshtools.j2ssh.authentication.TerminatedStateException",
      "uk.ac.sanger.artemis.io.EntryInformation",
      "htsjdk.samtools.TextTagCodec",
      "htsjdk.samtools.util.BlockCompressedInputStream",
      "org.biojava.bio.symbol.SimpleGeneticCodeTable",
      "uk.ac.sanger.artemis.sequence.MarkerChangeListener",
      "org.biojava.bio.seq.io.SeqIOListener",
      "org.biojava.utils.ChangeType",
      "uk.ac.sanger.artemis.io.ReadFormatException",
      "htsjdk.samtools.SRAIterator$RecordRangeInfo",
      "com.sshtools.j2ssh.sftp.SftpSubsystemClient",
      "com.sshtools.j2ssh.util.InvalidStateException",
      "htsjdk.variant.variantcontext.VariantContextComparator",
      "org.biojava.utils.ChangeListener",
      "htsjdk.samtools.PathInputResource",
      "uk.ac.sanger.artemis.components.BasePlot",
      "uk.ac.sanger.artemis.GotoEvent",
      "org.biojava.utils.AbstractChangeable",
      "org.gmod.schema.sequence.FeaturePub",
      "uk.ac.sanger.artemis.EntrySourceVector",
      "htsjdk.samtools.cram.io.CountingInputStream",
      "htsjdk.samtools.SamFlagField",
      "org.biojava.bio.symbol.EmptyAlphabet",
      "uk.ac.sanger.artemis.components.ViewMenu",
      "uk.ac.sanger.artemis.components.StickyFileChooser",
      "org.gmod.schema.dao.SchemaDaoI",
      "uk.ac.sanger.artemis.LastSegmentException",
      "uk.ac.sanger.artemis.ChangeEvent",
      "htsjdk.samtools.SamStreams",
      "org.biojava.bio.seq.SimpleFeatureRealizer",
      "uk.ac.sanger.artemis.io.GenbankStreamFeature",
      "uk.ac.sanger.artemis.util.OutOfRangeException",
      "uk.ac.sanger.artemis.components.alignment.CRAMReferenceSequenceFile",
      "htsjdk.samtools.util.Md5CalculatingOutputStream",
      "org.biojava.bio.symbol.SimpleBasisSymbol",
      "htsjdk.samtools.SamReader$PrimitiveSamReader",
      "org.biojava.utils.lsid.Identifiable",
      "org.biojava.bio.seq.io.CharacterTokenization$TPStreamParser",
      "org.biojava.bio.seq.DNATools$DNAComplementTranslationTable",
      "uk.ac.sanger.artemis.io.EMBLObject",
      "htsjdk.samtools.SAMFileSource",
      "uk.ac.sanger.artemis.ActionVector",
      "uk.ac.sanger.artemis.sequence.Strand",
      "uk.ac.sanger.artemis.chado.GmodDAO",
      "uk.ac.sanger.artemis.components.GotoMenu",
      "uk.ac.sanger.artemis.components.FeatureEdit",
      "org.apache.log4j.spi.DefaultRepositorySelector",
      "uk.ac.sanger.artemis.OptionChangeListener",
      "uk.ac.sanger.artemis.io.Sequence",
      "uk.ac.sanger.artemis.util.LargeObjectDocument",
      "htsjdk.samtools.SamInputResource",
      "htsjdk.samtools.SAMRecordCoordinateComparator",
      "uk.ac.sanger.artemis.io.GenbankMisc",
      "htsjdk.variant.vcf.VCFFormatHeaderLine",
      "uk.ac.sanger.artemis.chado.JdbcDAO",
      "org.biojava.utils.bytecode.GeneratedClassLoader",
      "uk.ac.sanger.artemis.io.Location",
      "uk.ac.sanger.artemis.components.EntryGroupDisplay",
      "org.biojava.utils.stax.StAXContentHandler",
      "uk.ac.sanger.artemis.components.SwingWorker",
      "uk.ac.sanger.artemis.util.RemoteFileDocument",
      "uk.ac.sanger.artemis.io.PublicDBStreamFeature",
      "htsjdk.samtools.Defaults",
      "org.apache.log4j.or.ObjectRenderer",
      "uk.ac.sanger.artemis.util.DatabaseLocationParser",
      "uk.ac.sanger.artemis.io.FastaStreamSequence",
      "uk.ac.sanger.artemis.io.FeatureEnumeration",
      "uk.ac.sanger.artemis.components.QualifierTextArea$2",
      "org.biojava.bio.symbol.AlphabetManager$WellKnownTokenizationWrapper",
      "uk.ac.sanger.artemis.components.CanvasPanel",
      "htsjdk.tribble.util.URLHelper",
      "org.biojava.utils.ChangeForwarder",
      "uk.ac.sanger.artemis.io.StreamFeature",
      "uk.ac.sanger.artemis.util.ReadOnlyException",
      "htsjdk.samtools.SRAInputResource",
      "htsjdk.samtools.SamFileValidator$PairEndInfoMap",
      "org.biojava.bio.symbol.GappedSymbolList",
      "uk.ac.sanger.artemis.components.GraphMenu",
      "htsjdk.samtools.util.BinaryCodec",
      "uk.ac.sanger.artemis.EntryChangeEvent",
      "uk.ac.sanger.artemis.components.PlotMouseListener",
      "org.biojava.bio.symbol.HashedAlphabetIndex$HashComparator",
      "org.biojava.bio.seq.FeatureFilter",
      "uk.ac.sanger.artemis.Feature",
      "uk.ac.sanger.artemis.io.EmblDocumentEntry",
      "org.biojava.utils.cache.KeyedWeakReference",
      "org.biojava.bio.seq.homol.HomologyFeature$Template",
      "htsjdk.samtools.BAMIteratorFilter",
      "uk.ac.sanger.artemis.components.SelectMenu",
      "com.sshtools.j2ssh.subsystem.SubsystemChannel",
      "uk.ac.sanger.artemis.io.DocumentFeature",
      "org.biojava.bio.SimpleAnnotation",
      "com.sshtools.j2ssh.FileTransferProgress",
      "org.biojava.bio.symbol.IntegerAlphabet",
      "uk.ac.sanger.artemis.plot.UserDataAlgorithm",
      "htsjdk.samtools.SAMFileHeader$PgIdGenerator",
      "uk.ac.sanger.artemis.plot.Algorithm",
      "org.biojava.bio.seq.impl.FeatureImpl",
      "org.apache.log4j.spi.LoggerFactory",
      "org.biojava.bio.symbol.DoubleAlphabet$DoubleSymbol",
      "org.apache.log4j.spi.Configurator",
      "htsjdk.variant.vcf.VCFInfoHeaderLine",
      "uk.ac.sanger.artemis.io.MSPcrunchStreamFeature",
      "org.biojava.bio.seq.impl.SimpleFramedFeature",
      "htsjdk.samtools.SamReaderFactory",
      "htsjdk.samtools.util.BlockGunzipper",
      "htsjdk.samtools.util.SequenceUtil",
      "org.gmod.schema.sequence.Feature",
      "org.biojava.bio.symbol.FiniteAlphabet",
      "uk.ac.sanger.artemis.components.SelectionSubMenu",
      "org.biojava.bio.seq.homol.SimilarityPairFeature$Template",
      "org.biojava.bio.seq.FeatureFilter$OnlyChildren",
      "org.apache.log4j.PropertyWatchdog",
      "uk.ac.sanger.artemis.io.KeyVector",
      "org.gmod.schema.sequence.FeatureProp",
      "htsjdk.samtools.cram.common.CramVersionPolicies",
      "org.biojava.bio.Annotatable",
      "org.gmod.schema.utils.Rankable",
      "uk.ac.sanger.artemis.components.FileViewer$2",
      "uk.ac.sanger.artemis.components.EntryGroupMenu",
      "htsjdk.samtools.seekablestream.SeekablePathStream",
      "uk.ac.sanger.artemis.FeatureVector",
      "uk.ac.sanger.artemis.io.EmblStreamSequence",
      "org.biojava.utils.stax.StringElementHandlerBase",
      "org.biojava.bio.seq.io.CrossProductTokenization",
      "htsjdk.samtools.SAMTextHeaderCodec$ParsedHeaderLine",
      "htsjdk.tribble.TribbleException$InternalCodecException",
      "org.biojava.bio.seq.FeatureHolder",
      "org.biojava.bio.symbol.FundamentalAtomicSymbol",
      "htsjdk.samtools.util.LineReader",
      "org.biojava.bio.symbol.AlphabetManager$AlphabetManagerHandler$CharacterTokenizationHandler$MappingHandler",
      "htsjdk.samtools.SAMValidationError$Type",
      "org.biojava.utils.ChangeForwarder$Retyper",
      "org.gmod.schema.dao.OrganismDaoI",
      "htsjdk.samtools.SRAFileReader",
      "htsjdk.samtools.util.BlockCompressedInputStream$FileTermination",
      "org.biojava.bio.seq.io.SymbolTokenization",
      "com.sshtools.j2ssh.session.SessionChannelClient",
      "uk.ac.sanger.artemis.FeaturePredicateConjunction",
      "htsjdk.samtools.util.LocationAware",
      "uk.ac.sanger.artemis.io.QualifierInfo",
      "uk.ac.sanger.artemis.io.IndexedGFFDocumentEntry$1",
      "org.biojava.utils.Unchangeable",
      "org.apache.log4j.Priority",
      "uk.ac.sanger.artemis.components.WriteMenu",
      "htsjdk.samtools.cram.structure.CramHeader",
      "uk.ac.sanger.artemis.components.IndexReferenceListener",
      "uk.ac.sanger.artemis.components.FeatureDisplay$3",
      "org.biojava.bio.seq.RNATools$RNAComplementTranslationTable",
      "org.apache.log4j.LogManager",
      "org.gmod.schema.sequence.FeatureLoc",
      "org.biojava.bio.symbol.AlphabetIndex",
      "org.biojava.bio.symbol.AlphabetManager$AlphabetManagerHandler$SymbolHandler",
      "uk.ac.sanger.artemis.components.FeatureDisplay$2",
      "org.biojava.bio.seq.FeatureHolder$EmptyFeatureHolder",
      "uk.ac.sanger.artemis.components.alignment.SnpPanel",
      "uk.ac.sanger.artemis.io.DocumentEntry",
      "org.biojava.bio.symbol.SimpleAtomicSymbol",
      "htsjdk.variant.vcf.VCFHeader",
      "org.biojava.bio.symbol.EmptySymbolList",
      "org.apache.log4j.DefaultCategoryFactory",
      "uk.ac.sanger.artemis.io.BlastStreamFeature",
      "uk.ac.sanger.artemis.components.alignment.ReadCountDialog",
      "htsjdk.tribble.util.ParsingUtils",
      "uk.ac.sanger.artemis.components.genebuilder.ortholog.MatchPanel",
      "org.apache.log4j.or.RendererMap",
      "htsjdk.samtools.cram.ref.CRAMReferenceSource",
      "uk.ac.sanger.artemis.io.OutOfDateException",
      "htsjdk.samtools.BAMRecord",
      "uk.ac.sanger.artemis.ExternalProgram",
      "com.sshtools.j2ssh.transport.Service",
      "htsjdk.samtools.util.StringUtil",
      "uk.ac.sanger.artemis.components.alignment.ReadGroupsFrame",
      "uk.ac.sanger.artemis.io.EmblMisc",
      "org.biojava.bio.symbol.AbstractAlphabet",
      "org.gmod.schema.general.DbXRef",
      "htsjdk.samtools.BamIndexValidator",
      "uk.ac.sanger.artemis.io.PartialSequence",
      "org.biojava.utils.ChangeListener$AlwaysVetoListener",
      "uk.ac.sanger.artemis.EntryGroupChangeEvent",
      "uk.ac.sanger.artemis.plot.ScaledChiAlgorithm",
      "htsjdk.samtools.SAMTagUtil",
      "uk.ac.sanger.artemis.Entry",
      "org.biojava.bio.symbol.AlphabetManager$SizeQueen",
      "org.biojava.bio.symbol.HashedAlphabetIndex",
      "org.biojava.utils.ClassTools",
      "uk.ac.sanger.artemis.io.EmblStreamFeature",
      "uk.ac.sanger.artemis.io.QualifierParseException",
      "org.biojava.utils.cache.WeakValueHashMap",
      "org.biojava.bio.symbol.IllegalSymbolException",
      "uk.ac.sanger.artemis.ExternalProgramVector",
      "htsjdk.samtools.SAMReadGroupRecord",
      "org.apache.log4j.CategoryKey",
      "uk.ac.sanger.artemis.plot.MRIAlgorithm",
      "htsjdk.samtools.SamReader$ReaderImplementation",
      "uk.ac.sanger.artemis.io.Qualifier",
      "uk.ac.sanger.artemis.util.InputStreamProgressListener",
      "htsjdk.samtools.UrlInputResource",
      "uk.ac.sanger.artemis.components.alignment.PopupMessageFrame",
      "org.biojava.bio.seq.RemoteFeature",
      "htsjdk.samtools.seekablestream.SeekableBufferedStream",
      "htsjdk.samtools.CRAMFileReader$1",
      "org.apache.log4j.helpers.OnlyOnceErrorHandler",
      "htsjdk.samtools.seekablestream.SeekableHTTPStream",
      "org.biojava.bio.symbol.TranslationTable",
      "uk.ac.sanger.artemis.io.DocumentEntryAutosaveThread",
      "uk.ac.sanger.artemis.components.EntryGroupPanel",
      "org.biojava.bio.symbol.CrossProductAlphabetIndex",
      "com.sshtools.j2ssh.sftp.FileAttributes",
      "org.biojava.bio.seq.impl.SimpleStrandedFeature",
      "uk.ac.sanger.artemis.sequence.Marker",
      "org.biojava.bio.seq.Feature$ByLocationComparator",
      "uk.ac.sanger.artemis.plot.CSCSAlgorithm",
      "org.biojava.bio.symbol.DoubleAlphabet",
      "org.apache.log4j.ProvisionNode",
      "com.sshtools.j2ssh.authentication.SshAuthenticationClient",
      "uk.ac.sanger.artemis.io.GFFDocumentEntry",
      "org.biojava.bio.symbol.AbstractSymbolList$EditTranslater",
      "uk.ac.sanger.artemis.io.GFFStreamFeature",
      "htsjdk.samtools.AbstractBAMFileIndex$IndexFileBuffer",
      "org.biojava.bio.symbol.AlphabetManager$AlphabetManagerHandler",
      "org.biojava.bio.symbol.AtomicSymbol",
      "com.sshtools.j2ssh.transport.MessageStoreEOFException",
      "org.biojava.bio.dist.Distribution",
      "uk.ac.sanger.artemis.components.QualifierChoice",
      "org.apache.log4j.spi.RootLogger",
      "htsjdk.samtools.cram.structure.SliceIO",
      "org.biojava.bio.BioError",
      "uk.ac.sanger.artemis.EntryVector",
      "htsjdk.variant.vcf.VCFContigHeaderLine",
      "org.apache.log4j.spi.ErrorHandler",
      "htsjdk.samtools.util.AsyncBlockCompressedInputStream",
      "uk.ac.sanger.artemis.ExternalProgramException",
      "uk.ac.sanger.artemis.io.QualifierVector",
      "uk.ac.sanger.artemis.components.alignment.BamView$Ruler",
      "org.apache.log4j.spi.RendererSupport",
      "org.biojava.bio.seq.OptimizableFilter",
      "uk.ac.sanger.artemis.FeatureChangeListener",
      "uk.ac.sanger.artemis.FeatureChangeEvent",
      "uk.ac.sanger.artemis.util.FileDocument",
      "htsjdk.samtools.util.Log",
      "uk.ac.sanger.artemis.components.alignment.GroupBamFrame",
      "uk.ac.sanger.artemis.plot.CodonWeight",
      "htsjdk.samtools.util.Iso8601Date$1",
      "org.biojava.bio.symbol.AlphabetManager$GapSymbol",
      "uk.ac.sanger.artemis.io.LocationParseNodeVector",
      "uk.ac.sanger.artemis.io.LineGroup",
      "uk.ac.sanger.artemis.Entry$FeatureEnumerator",
      "uk.ac.sanger.artemis.GotoEventSource",
      "org.apache.commons.net.ftp.FTPConnectionClosedException",
      "htsjdk.variant.variantcontext.Allele",
      "htsjdk.samtools.util.IOUtil",
      "org.biojava.utils.ChangeSupport",
      "org.biojava.bio.seq.io.AlternateTokenization",
      "org.biojava.bio.molbio.RestrictionSite",
      "uk.ac.sanger.artemis.io.IndexedGFFDocumentEntry",
      "org.biojava.bio.seq.StrandedFeature$Strand",
      "uk.ac.sanger.artemis.components.SequenceComboBox",
      "uk.ac.sanger.artemis.components.genebuilder.ortholog.MatchPanel$3",
      "htsjdk.samtools.SAMFileSpan",
      "uk.ac.sanger.artemis.plot.GCSDWindowAlgorithm",
      "org.biojava.bio.symbol.AlphabetManager$1",
      "org.apache.log4j.Logger",
      "uk.ac.sanger.artemis.FeatureSegment",
      "org.biojava.bio.seq.homol.HomologyFeature",
      "uk.ac.sanger.artemis.EntryGroup",
      "uk.ac.sanger.artemis.io.Feature",
      "org.biojava.bio.AbstractAnnotation",
      "org.biojava.bio.symbol.ManyToOneTranslationTable",
      "com.sshtools.j2ssh.authentication.AuthenticationProtocolListener",
      "org.apache.log4j.helpers.LogLog",
      "uk.ac.sanger.artemis.io.QualifierInfoVector",
      "htsjdk.samtools.SAMBinaryTagAndUnsignedArrayValue",
      "htsjdk.samtools.BamIndexValidator$IndexValidationStringency",
      "htsjdk.samtools.cram.structure.Slice",
      "htsjdk.samtools.cram.common.CramVersions",
      "uk.ac.sanger.artemis.components.DisplayComponent",
      "org.apache.log4j.spi.RepositorySelector",
      "org.biojava.utils.ListTools$SeriesList",
      "org.biojava.bio.symbol.SimpleSymbolList",
      "uk.ac.sanger.artemis.components.EntryEdit$8",
      "uk.ac.sanger.artemis.plot.AGWindowAlgorithm",
      "uk.ac.sanger.artemis.components.EntryEdit$7",
      "uk.ac.sanger.artemis.components.alignment.BamViewRecord",
      "uk.ac.sanger.artemis.components.alignment.LookSeqPanel",
      "org.biojava.utils.ChangeListener$LoggingListener",
      "org.biojava.bio.symbol.SingletonAlphabet",
      "uk.ac.sanger.artemis.io.ReadEvent",
      "uk.ac.sanger.artemis.components.variant.GraphPanel",
      "org.biojava.bio.SmallAnnotation",
      "htsjdk.samtools.SamFlagField$3",
      "htsjdk.samtools.SamFlagField$4",
      "org.biojava.bio.symbol.ReversibleTranslationTable",
      "uk.ac.sanger.artemis.io.Entry",
      "htsjdk.samtools.SamFlagField$5",
      "uk.ac.sanger.artemis.util.LinePushBackReader",
      "org.apache.log4j.or.DefaultRenderer",
      "uk.ac.sanger.artemis.util.ProgressInputStream",
      "htsjdk.samtools.SamReader$AssertingIterator",
      "htsjdk.samtools.SamFlagField$1",
      "htsjdk.samtools.SamFlagField$2",
      "htsjdk.variant.vcf.VCFHeaderLine",
      "htsjdk.samtools.cram.build.CramIO",
      "uk.ac.sanger.artemis.io.FeatureVector",
      "uk.ac.sanger.artemis.chado.ChadoTransactionManager",
      "uk.ac.sanger.artemis.components.alignment.BamView",
      "htsjdk.samtools.SAMFormatException",
      "org.biojava.bio.symbol.LinearAlphabetIndex",
      "org.biojava.bio.symbol.AlphabetManager$AlphabetManagerHandler$SymbolHandler$1",
      "htsjdk.samtools.AbstractBAMFileIndex$RandomAccessFileBuffer",
      "htsjdk.samtools.seekablestream.SeekableStreamFactory",
      "org.gmod.schema.organism.Organism",
      "org.apache.log4j.Appender",
      "htsjdk.samtools.cram.structure.BlockCompressionMethod",
      "org.biojava.bio.symbol.SimpleReversibleTranslationTable",
      "htsjdk.samtools.InputResource$Type",
      "org.biojava.bio.symbol.Alphabet",
      "org.biojava.bio.symbol.AbstractSymbolList$EditScreener",
      "org.biojava.utils.ListTools$5",
      "org.biojava.bio.seq.Frame",
      "uk.ac.sanger.artemis.chado.ChadoTransaction",
      "com.sshtools.j2ssh.transport.SshMsgServiceAccept",
      "org.biojava.utils.AssertionFailure",
      "uk.ac.sanger.artemis.FeaturePredicate",
      "com.sshtools.j2ssh.authentication.AuthenticationProtocolException",
      "htsjdk.samtools.SAMValidationError",
      "uk.ac.sanger.artemis.Logger",
      "org.biojava.bio.seq.IsTopLevel",
      "uk.ac.sanger.artemis.components.alignment.SAMRecordFilter",
      "org.gmod.schema.sequence.FeatureSynonym",
      "org.apache.log4j.spi.LoggingEvent",
      "org.apache.commons.compress.utils.CountingOutputStream",
      "org.biojava.bio.symbol.AlphabetManager$WellKnownGapSymbol",
      "htsjdk.samtools.SamReader",
      "htsjdk.samtools.sra.SRAAccession",
      "uk.ac.sanger.artemis.io.QualifierLazyLoading",
      "uk.ac.sanger.artemis.util.DatabaseDocument",
      "uk.ac.sanger.artemis.components.EntryEdit",
      "uk.ac.sanger.artemis.FeatureEnumeration",
      "uk.ac.sanger.artemis.components.FeatureList",
      "org.biojava.bio.symbol.AbstractSymbolList",
      "uk.ac.sanger.artemis.io.StreamFeatureTable",
      "uk.ac.sanger.artemis.SelectionChangeListener",
      "org.gmod.schema.dao.PubDaoI",
      "uk.ac.sanger.artemis.GotoListener"
    );
  } 
  private static void initMocksToAvoidTimeoutsInTheTests() throws ClassNotFoundException { 
    mock(Class.forName("java.net.URLStreamHandler", false, BamUtils_ESTest_scaffolding.class.getClassLoader()));
  }

  private static void resetClasses() {
    org.evosuite.runtime.classhandling.ClassResetter.getInstance().setClassLoader(BamUtils_ESTest_scaffolding.class.getClassLoader()); 

    org.evosuite.runtime.classhandling.ClassStateSupport.resetClasses(
      "org.apache.log4j.Category",
      "org.apache.log4j.Logger",
      "org.apache.log4j.Priority",
      "org.apache.log4j.Level",
      "org.apache.log4j.or.RendererMap",
      "org.apache.log4j.helpers.OptionConverter",
      "org.apache.log4j.helpers.Loader",
      "org.apache.log4j.helpers.LogLog",
      "org.apache.log4j.PropertyConfigurator",
      "org.apache.log4j.LogManager",
      "org.apache.log4j.CategoryKey",
      "org.apache.log4j.ProvisionNode",
      "uk.ac.sanger.artemis.components.alignment.BamUtils",
      "uk.ac.sanger.artemis.components.alignment.ReadCountDialog",
      "uk.ac.sanger.artemis.util.DatabaseDocument",
      "uk.ac.sanger.artemis.util.StringVector",
      "uk.ac.sanger.artemis.components.Splash",
      "uk.ac.sanger.artemis.util.FastVector",
      "uk.ac.sanger.artemis.io.KeyVector",
      "uk.ac.sanger.artemis.io.Key",
      "uk.ac.sanger.artemis.io.QualifierInfo",
      "uk.ac.sanger.artemis.ExternalProgram",
      "uk.ac.sanger.artemis.Options",
      "htsjdk.samtools.ValidationStringency",
      "htsjdk.samtools.SamReaderFactory$Option$1",
      "htsjdk.samtools.SamReaderFactory$Option$2",
      "htsjdk.samtools.SamReaderFactory$Option$3",
      "htsjdk.samtools.SamReaderFactory$Option$4",
      "htsjdk.samtools.SamReaderFactory$Option$5",
      "htsjdk.samtools.util.Log$LogLevel",
      "htsjdk.samtools.util.Log",
      "htsjdk.samtools.SamReaderFactory$Option",
      "htsjdk.samtools.SamReaderFactory$SamReaderFactoryImpl",
      "htsjdk.samtools.DefaultSAMRecordFactory",
      "htsjdk.samtools.util.BlockGunzipper",
      "htsjdk.samtools.SamFlagField$1",
      "htsjdk.samtools.SamFlagField$2",
      "htsjdk.samtools.SamFlagField$3",
      "htsjdk.samtools.SamFlagField$4",
      "htsjdk.samtools.SamFlagField$5",
      "htsjdk.samtools.SamFlagField",
      "htsjdk.samtools.Defaults",
      "htsjdk.samtools.CustomReaderFactory",
      "htsjdk.samtools.SamReaderFactory",
      "htsjdk.samtools.SamReader$Type",
      "htsjdk.samtools.SAMFileHeader$SortOrder",
      "htsjdk.samtools.cram.structure.CramHeader",
      "htsjdk.samtools.cram.build.CramIO",
      "htsjdk.samtools.cram.common.CramVersions",
      "htsjdk.samtools.cram.structure.BlockCompressionMethod",
      "htsjdk.samtools.cram.structure.BlockContentType",
      "htsjdk.samtools.SAMValidationError$Severity",
      "htsjdk.samtools.SAMValidationError$Type",
      "htsjdk.samtools.cram.structure.SliceIO",
      "htsjdk.samtools.SAMTagUtil",
      "htsjdk.samtools.cram.structure.SliceIO$1",
      "htsjdk.samtools.cram.structure.ContainerIO",
      "htsjdk.samtools.cram.common.CramVersionPolicies",
      "htsjdk.samtools.seekablestream.SeekableStreamFactory",
      "htsjdk.samtools.BamIndexValidator$IndexValidationStringency",
      "uk.ac.sanger.artemis.io.LineGroup",
      "uk.ac.sanger.artemis.io.Packing",
      "uk.ac.sanger.artemis.sequence.Bases",
      "uk.ac.sanger.artemis.sequence.Strand",
      "htsjdk.samtools.SAMException",
      "htsjdk.samtools.InputResource$Type",
      "htsjdk.samtools.SamFiles",
      "htsjdk.samtools.seekablestream.SeekableFileStream",
      "htsjdk.samtools.util.RuntimeIOException",
      "uk.ac.sanger.artemis.util.OutOfRangeException",
      "uk.ac.sanger.artemis.components.EntryEdit",
      "uk.ac.sanger.artemis.components.BasePlotGroup",
      "uk.ac.sanger.artemis.components.Plot",
      "uk.ac.sanger.artemis.components.BasePlot",
      "uk.ac.sanger.artemis.io.PublicDBDocumentEntry",
      "uk.ac.sanger.artemis.io.FeatureVector",
      "uk.ac.sanger.artemis.ChangeEvent",
      "uk.ac.sanger.artemis.EntryGroupChangeEvent",
      "uk.ac.sanger.artemis.io.QualifierVector",
      "uk.ac.sanger.artemis.io.SimpleDocumentFeature",
      "uk.ac.sanger.artemis.io.PublicDBStreamFeature",
      "uk.ac.sanger.artemis.io.ReadEvent",
      "uk.ac.sanger.artemis.io.EntryInformationException",
      "uk.ac.sanger.artemis.io.InvalidKeyException",
      "htsjdk.samtools.AbstractSAMHeaderRecord",
      "htsjdk.samtools.SAMFileHeader",
      "htsjdk.samtools.util.IOUtil",
      "htsjdk.samtools.SamFileValidator",
      "htsjdk.samtools.util.Histogram",
      "htsjdk.samtools.CRAMFileReader",
      "htsjdk.samtools.cram.ref.ReferenceSource",
      "uk.ac.sanger.artemis.components.filetree.FileList",
      "uk.ac.sanger.artemis.j2ssh.SshLogin",
      "htsjdk.samtools.seekablestream.SeekablePathStream",
      "org.biojava.utils.ChangeType",
      "org.biojava.bio.seq.Feature",
      "htsjdk.samtools.reference.ReferenceSequenceFileFactory$1",
      "htsjdk.samtools.reference.ReferenceSequenceFileFactory",
      "uk.ac.sanger.artemis.FeaturePredicateConjunction",
      "htsjdk.samtools.sra.SRAAccession",
      "htsjdk.samtools.SRAFileReader",
      "htsjdk.samtools.SAMTag",
      "htsjdk.samtools.SAMRecord",
      "org.biojava.bio.AbstractAnnotation",
      "org.biojava.bio.SimpleAnnotation",
      "org.biojava.bio.seq.SimpleFeatureRealizer",
      "org.biojava.bio.seq.impl.FeatureImpl$1",
      "org.biojava.bio.seq.impl.FeatureImpl",
      "org.biojava.bio.symbol.AlphabetManager$GapSymbol",
      "org.biojava.bio.symbol.AlphabetManager$SizeQueen",
      "org.biojava.bio.symbol.IntegerAlphabet",
      "org.biojava.bio.symbol.DoubleAlphabet",
      "org.biojava.bio.symbol.SimpleSymbol",
      "org.biojava.bio.symbol.AbstractSimpleBasisSymbol",
      "org.biojava.bio.symbol.AlphabetManager$WellKnownGapSymbol",
      "org.biojava.bio.EmptyAnnotation",
      "org.biojava.bio.Annotation",
      "org.biojava.bio.symbol.EmptyAlphabet",
      "org.biojava.bio.symbol.Alphabet",
      "org.biojava.utils.ListTools",
      "org.biojava.utils.SingletonList",
      "org.biojava.bio.SmallAnnotation",
      "org.biojava.utils.lsid.LifeScienceIdentifier",
      "org.biojava.utils.SmallMap",
      "org.biojava.bio.symbol.AlphabetManager$WellKnownBasisSymbol",
      "org.biojava.bio.symbol.AlphabetManager$WellKnownAtomicSymbol",
      "org.biojava.bio.symbol.FundamentalAtomicSymbol",
      "org.biojava.utils.ChangeListener",
      "org.biojava.utils.ChangeSupport",
      "org.biojava.bio.Annotatable",
      "org.biojava.bio.symbol.AbstractAlphabet",
      "org.biojava.bio.symbol.SingletonAlphabet",
      "org.biojava.bio.symbol.SimpleAlphabet",
      "org.biojava.bio.symbol.AlphabetManager$WellKnownAlphabet",
      "org.biojava.bio.symbol.AlphabetManager$ImmutableWellKnownAlphabetWrapper",
      "org.biojava.bio.seq.io.CharacterTokenization",
      "org.biojava.bio.symbol.SimpleBasisSymbol",
      "org.biojava.bio.seq.io.AlternateTokenization",
      "org.biojava.bio.symbol.AlphabetManager",
      "org.biojava.bio.symbol.AlphabetManager$WellKnownTokenizationWrapper",
      "org.biojava.bio.symbol.SimpleSymbolList",
      "org.biojava.bio.seq.io.SymbolTokenization",
      "org.biojava.bio.symbol.SimpleReversibleTranslationTable",
      "org.biojava.bio.symbol.SimpleCrossProductAlphabet",
      "org.biojava.utils.ListTools$Triplet",
      "org.biojava.bio.symbol.SimpleAtomicSymbol",
      "org.biojava.bio.seq.io.WordTokenization",
      "org.biojava.bio.seq.io.NameTokenization",
      "org.biojava.bio.symbol.SimpleManyToOneTranslationTable",
      "org.biojava.bio.symbol.SimpleGeneticCodeTable",
      "org.biojava.bio.symbol.AbstractSymbolList$ListView",
      "org.biojava.bio.BioException",
      "org.biojava.bio.symbol.IllegalSymbolException",
      "org.biojava.bio.BioError",
      "org.biojava.bio.seq.RNATools",
      "org.biojava.bio.seq.DNATools",
      "uk.ac.sanger.artemis.sequence.AminoAcidSequence",
      "htsjdk.samtools.SAMSequenceDictionary",
      "htsjdk.variant.vcf.VCFHeader",
      "htsjdk.tribble.util.ParsingUtils",
      "uk.ac.sanger.artemis.io.LocationParseNodeVector",
      "uk.ac.sanger.artemis.io.ReadFormatException",
      "uk.ac.sanger.artemis.io.LocationParseException",
      "uk.ac.sanger.artemis.io.RangeVector",
      "uk.ac.sanger.artemis.io.LocationParseNode",
      "uk.ac.sanger.artemis.EntryChangeEvent",
      "uk.ac.sanger.artemis.components.alignment.BamView",
      "com.sshtools.j2ssh.sftp.FileAttributes",
      "htsjdk.samtools.util.StringUtil",
      "htsjdk.variant.variantcontext.Allele",
      "uk.ac.sanger.artemis.components.CanvasPanel",
      "uk.ac.sanger.artemis.components.EntryGroupPanel",
      "uk.ac.sanger.artemis.components.FeatureDisplay",
      "uk.ac.sanger.artemis.components.ZoomScrollBar",
      "uk.ac.sanger.artemis.components.DisplayAdjustmentEvent",
      "org.apache.log4j.spi.Filter",
      "htsjdk.samtools.util.SequenceUtil",
      "htsjdk.samtools.SAMFlag",
      "uk.ac.sanger.artemis.components.genebuilder.ortholog.MatchPanel",
      "uk.ac.sanger.artemis.chado.ChadoTransactionManager",
      "uk.ac.sanger.artemis.plot.UserDataAlgorithm",
      "htsjdk.samtools.util.BlockCompressedInputStream",
      "htsjdk.samtools.SAMTextHeaderCodec$HeaderRecordType",
      "htsjdk.samtools.SAMTextHeaderCodec",
      "htsjdk.samtools.TextTagCodec",
      "org.emboss.jemboss.JembossParams",
      "uk.ac.sanger.artemis.util.DatabaseLocationParser",
      "org.apache.log4j.MDC",
      "htsjdk.samtools.SAMSequenceRecord",
      "uk.ac.sanger.artemis.io.FeatureTree",
      "uk.ac.sanger.artemis.io.DocumentEntryAutosaveThread",
      "uk.ac.sanger.artemis.FeatureSegment"
    );
  }
}
